tmt/steps/report/junit.py

```diff
@@ -58,40 +67,230 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
-        except (IndexError, AttributeError):
-            main_log = None
-        case = junit_xml.TestCase(
-            result.name,
-            classname=None,
-            elapsed_sec=duration_to_seconds(result.duration))
-
-        if report.data.include_output_log:
-            case.stdout = main_log
-
-        # Map tmt OUTCOME to JUnit states
-        if result.result == tmt.result.ResultOutcome.ERROR:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.FAIL:
-            case.add_failure_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.INFO:
-            case.add_skipped_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.WARN:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        # Passed state is the default
-        suite.test_cases.append(case)
-
-    return suite
+            # S320: Parsing of untrusted data is known to be vulnerable to XML
+            # attacks.
+            tree_root: XMLElement = etree.fromstring(xml_data, xml_parser)  # noqa: S320
+
+        except etree.XMLSyntaxError as e:
+            phase.warn(
+                'The generated XML output is not a valid XML file or it is not valid against the '
+                'XSD schema.')
+
+            if flavor != CUSTOM_FLAVOR_NAME:
+                phase.warn('Please, report this problem to project maintainers.')
+
+            for err in e.error_log:
+                phase.warn(str(err))
+
+            # Return the prettified XML without checking the XSD
+            del xml_parser_kwargs['schema']
+
+            try:
+                tree_root = etree.fromstring(  # noqa: S320
+                    xml_data, etree.XMLParser(**xml_parser_kwargs))
+            except etree.XMLSyntaxError as e:
+                phase.warn(str(e))
+                phase.verbose('data', xml_data, 'red')
+                raise tmt.utils.ReportError(
+                    'The generated XML output is not a valid XML file. Use `--verbose` argument '
+                    'to show the output.')
+
+        # Do not be fooled by the `encoding` parameter: even with `utf-8`,
+        # `tostring()` will still return bytes. `unicode`, on the other
+        # hand, would give us a string, but then, no XML declaration.
+        # So, we get bytes, and we need to apply `decode()` on our own.
+        xml_output: bytes = etree.tostring(
+            tree_root,
+            xml_declaration=True,
+            pretty_print=prettify,
+            # The 'utf-8' encoding must be used instead of 'unicode', otherwise
+            # the XML declaration is not included in the output.
+            encoding='utf-8')
+
+        return str(xml_output.decode('utf-8'))
+
+    except ImportError:
```

Comment:
Hmm, how about wrapping only the import into this `try-except` block? That way the rest of the code which imports nothing could save one level of pointless indentation.

Reply:
Good catch. Thanks.